### PR TITLE
RFC new proposal system

### PR DESCRIPTION
**Pull request by msutherl**
## Changes in this PR

- Supersedes [UP 0](https://github.com/urbit/proposals/blob/97cafe4f57b0053014e49a12076736a70a02c7b2/0.udon)
- Adds a proposal template
- Updates the README
- Converts all old proposals to the new format proposed in UP 0 (RFC)
- Converts all old comments to markdown files and renames their folders

## Abstract

This PR implements a new proposal system based on [COSS](https://rfc.unprotocols.org/spec:2/COSS/), [EIP](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1.md), [BIP](https://github.com/bitcoin/bips/blob/master/bip-0001.mediawiki), and [PEP](https://github.com/python/peps/blob/master/pep-0001.txt). UP 0 has been rewritten to formalize this system, and the remaining published UPs have been updated to match the format.

This system involves a public developer mailing list, dev@urbit.org, which can be found here: https://groups.google.com/a/urbit.org/forum/#!forum/dev.

The proposal system relies on @msutherl and @jtobin to carry out editorial duties described in UP 0.

## Motivation

Since the advent of [the original UP 0](https://github.com/urbit/proposals/blob/97cafe4f57b0053014e49a12076736a70a02c7b2/0.udon) by ~tonlur-sarret, Urbit's core developers have loosely followed this format to organize their work. For lack of a clear process for vetting and publishing proposals, however, most proposals since Fora was taken down and UP 5 was published have been circulated privately. This makes the Urbit's development more opaque to the world than it needs to be and deprives the project of productive community engagement.

## Rationale

In designing this system, we looked closely at the three dominant proposal system families in common use:

- IETF [RFCs](https://www.ietf.org/standards/rfcs/), used to create Internet standards
- [COSS](https://rfc.unprotocols.org/spec:2/COSS/), from the ZeroMQ project's C4 Collective Code Construction Contract
- Python [PEPs](https://www.python.org/dev/peps/) and their many derivatives, including Bitcoin's  [BIPs](https://github.com/bitcoin/bips/blob/master/bip-0001.mediawiki).

Takeaways from studying these systems:

1. The importance of the details is hard to predict in advance. Each system is an attempt to formalize a set of processes which are mostly implicit and evolve organically with the project and the specific people involved.
2. BIP and EIP were the closest to what we think might work for the Urbit project, however we were inspired by the simplicity of COSS.
3. Urbit has some unique requirements that require significant modifications to both the PEP and COSS structure, namely that we don't have a robust notion of "standards track" and there is a wider variety of proposal types given that we're building more than a programming language (PEP), distributed database (BIP, EIP), or protocol (COSS).

Bearing this in mind, the proposed system mostly copies various parts of these systems wholesale while attempting as much as possible to simplify, leave things open-ended, and accommodate a wider variety of proposal types.

We expect that what we've come up here will not perfectly match the patterns of use that will emerge and plan to evolve the documentation to match reality.

Some other notes:

- Markdown is used for familiarity and easy inclusion in urbit.org / docs
- TOML front-matter is used because our current website is being re-implemented as a Zola static-site, and Zola enforces TOML front-matter. Unfortunately, TOML isn't rendered well on Github (I emailed them about fixing this). This isn't set in stone.

## RFC

Feedback on all aspects of this PR is welcome. A few specific things I want to point out:

### Mailing List

- This system proposes using a mailing list for RFC review, following Bitcoin. I confirmed this with @philipcmonk, @jtobin, @vvisigoth, @galenwp and others in conversation, but if you have thoughts about this, please let me know. The proposed mailing list lives at https://groups.google.com/a/urbit.org/forum/#!forum/dev.

### Process

- Comments from old fora are archived in a comments folder. This is a bit of a wart, so just calling it out.
- I did not propose a method of including external assets in proposals. Currently, UP 8 makes reference to an image stored at media.urbit.org. This is probably not a good pattern. I'm deliberately leaving this open until our website infrastructure is sorted out.
- This PR replaces UP 0 rather than superseding it, which I thought was cleaner even though it breaks the rules. Also UP 2 references UP 0, so will need to note the exception and make reference to the original UP 0 in git history.
- It might make sense to explicitly encourage people to submit RFCs via GH issues in this repository, similar to [how Ethereum does it](https://github.com/ethereum/EIPs/issues).

### Proposals

- I'm not sure if I'm completely satisfied with the proposal type taxonomy: Feature, Architecture, Protocol, Informational, Process. On one hand, it feels overkill. On the other hand, it could be more specific. For instance, how should changes to Hoon be classified? And what is the essential difference between an architectural proposal and a feature proposal? The EIP system has fewer types and adds an additional category tag – I could imagine this making sense, but decided against adding additional structure before we see how these things are used.
- I'm also not completely certain about the workflow, which is based mostly on PEP, but simplified: RFC, Draft, Stable, Deprecated, Rejected, Deferred, Withdrawn. "Active" is omitted, and "Superseded" is a state of "Deferred".
- We had previously encouraged an Overview, Rationale, Specification, Implementation Plan structure. I'm proposing a more complex structure based on BIP/EIP that separates Motivation and Rationale and adds an Abstract. I think this is a better structure, but open to critiques.

## Next Steps

Previously published UPs are part of this PR, but many proposals exist that could be added here. Next step would be to identify these and fast-track them through the process.

### CODE REVIEW

**Comment by galenwp** [comment]
At first glance, lgtm. V happy to have something new here.

File: 000-purpose-and-guidelines.md [hunk 1]
```diff
@@ -65,8 +65,8 @@ requires = [
   "UP <number>"
   "UP <number>"
 ] 
-superceded-by = "UP <number>"
-supercedes = "UP <number>" 
+superseded-by = "UP <number>"
```
Comment (jtobin) [inline 000-purpose-and-guidelines.md]: here here

**Comment by jtobin** [comment]
@msutherl I'll give this a proper look-over a little later today!  Have some errands to run shortly.

File: 000-purpose-and-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+The author may encourage ongoing discussion in the PR comments or at the `comments` URI. UPs should remain as drafts until implemented, deferred, or rejected.
+
+### Stable 
+
+When a UP has been implemented, a PR may be submitted to change the `status` to "stable." This indicates that the UP describes functionality or a process that is currently in-use.
+
+### Deprecated
+
+A UP that describes functionality or a process that has been removed, fallen out of use, or superseded by another UP may be marked "deprecated." If superseded,
```
Comment (philipcmonk) [inline 000-purpose-and-guidelines.md]: unfinished sentence

Comment (msutherl) [reply 000-purpose-and-guidelines.md]: thanks – fixed!

**Comment by jtobin** [comment]
(We may want to revisit using Google Groups for the list at some point -- I didn't realise that it's kind of a hack to subscribe from a non-Google account.  Not a big deal, but a small deal, perhaps.)

**Comment by msutherl** [comment]
> (We may want to revisit using Google Groups for the list at some point -- I didn't realise that it's kind of a hack to subscribe from a non-Google account. Not a big deal, but a small deal, perhaps.)

This is indeed annoying. Perhaps we can write up a quick document about how to join with a non-Google email address. Looks like it requires in some cases setting up your email address as an "alternate" to your gmail address.

**Pull request merged**